Language extensions

feiranwang opened the pull request:
1. Support limit
2. Support plpy extractor
3. Support factor functions and, or, equal
4. Unify atom and condition
5. Add atom modifier
6. Support placeholder
7. Bug fix for aggregation function in nested expressions

feiranwang committed:
Add support for limit

feiranwang committed:
Add support for plpy extractor

feiranwang committed:
Add support for specifying factor functions. Change semantics specification

feiranwang committed:
Add support for IN, EXISTS, ANY, ALL

feiranwang committed:
Add support for left outer join

feiranwang committed:
Add support for binding weight to a constant

feiranwang committed:
Fix aggregation in nested expressions

feiranwang committed:
Fix outer join

feiranwang committed:
Unify atom and condition

feiranwang committed:
Remove trailing whitespaces

feiranwang committed:
Refactor compiling expression and alias

feiranwang committed:
Add support for atom modifier. Support EXIST

feiranwang committed:
Add support for unified body in modifier atom

feiranwang committed:
Add suport for not exist

feiranwang committed:
Add support for outer join modifier

feiranwang committed:
Support placeholder

feiranwang committed:
Merge master

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -66,10 +66,10 @@ import org.deepdive.ddlog.DeepDiveLog.Mode._
+  // NoAlias => use column name for dd_new_ tables
+  // WeightAlias => use alias for unknown weight in inference
+  // HeadAlias => use column name and alias for normal extraction rule
+  val NoAlias, WeightAlias, HeadAlias = Value
```
Thanks for renaming.  Much clearer!

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -176,124 +176,194 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
 
   // resolve an expression
-  def compileExpr(e: Expr, cq: ConjunctiveQuery, alias: AliasStyle, index : Int, isHead: Boolean) : String = {
+  def compileExpr(e: Expr, cq: ConjunctiveQuery, generateAlias: Expr => String = e => "") = {
```
I think taking a substitution function of type `String => String` (that corresponds to `compileVariable(_, cq)`) as the only extra parameter makes more sense here because it'll allows dropping `cq`, which is never used elsewhere.

Since `generateAlias` isn't doing anything inside `compileExprInner`, we should also take that completely out to the callers.  If you look at how `generateAlias` is being instantiated, it's all some form of `compileAlias`.  If that's the case, `compileAlias` can actually call `compileExpr`, appending the `" AS ..."`, and `compileExpr` can be as simple as `(Expr, String => String) => String`.

Could you revise the refactoring if these make sense?

netj replied:
After another look, I think it may make more sense to move `compileExpr`, `compileCond`, `generateSQLBody` etc. to `QuerySchema` class, and have implicit access to the `cq` they belong to, instead of having every function toss `cq` around and instantiate `QuerySchema` with it multiple times.

feiranwang replied:
For the first comment, I agree we can separate `generateAlias`.  For `compileVariable` I think it's not better than passing `cq`. It makes calling `compileExpr` harder, as instead of passing just a `cq`, now you pass another function that takes `cq` as an argument, e.g., compare `compileExpr(e, cq)` with `compileExpr(e, compileVariable(_,cq))`. This holds for every call site. 

For the second comment, I think it makes sense to move those compilation functions into `QuerySchema` and rename that class. That class is then responsible for compiling a single conjunctive query. This will solve the above problem I think.

netj replied:
I agree with your first point.  My suggestion would make each call awkward.  So, let's turn `QuerySchema` into a `DatalogRuleCompilation` or so that takes care of compiling each rule.  Can we also integrate the head compilation as well into this class?  I think that makes sense because the aliasing in the head has to rely on the same naming scheme used there, right?

feiranwang replied:
Sounds good. I'll also look into head compilation.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -176,124 +176,194 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
-    val conditionStr = z.conditions(0) map (compileCond(_, z))
+    val whereClause = generateWhereClause(z.bodies(0), "")
+    val whereClauseStr = if (whereClause.isEmpty) "" else "WHERE " + whereClause
+    val fromClauseStr = generateFromClause(z.bodies(0), "")
```
Can we inline these at the end where the whole SQL query is put together?

I like the structure we're getting at: defining functions that take care of a particular aspect of the compilation, then using them at the end to generate the final result.  So, let's move most `val`s down.

Defining a function that surrounds certain prefix/postfix strings only for non-empty args may be useful:

``` scala
def optionalClause(prefix: String, clause: String, postfix: String = ""): String =
  if (clause isEmpty) "" else prefix + clause + postfix
```

feiranwang replied:
Sounds good.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -176,124 +176,194 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
-    val groupbyStr = if (groupbyTerms.size == z.head.terms.size) {
+    // handle group by
+    // map head terms, leaving out aggregation functions
+    val groupbyTerms = z.head.terms.zipWithIndex flatMap { case (expr, index) =>
```
Why do you `zipWithIndex`?  `parititon` or `filter` may be useful now since you nicely defined a boolean predicate.  You can chain that with a `map compileExpr`

feiranwang replied:
zipWithIndex is redundant. filter looks good here.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -176,124 +176,194 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
-    val whereClauseConds = whereClause ++ conditionStr
-    val whereClauseStr = if (whereClauseConds isEmpty) "" else whereClauseConds.mkString("WHERE ", " AND ", "")
+    // limit clause
+    val limitStr = z.limit match {
```
Let's try to use the `optionalClause` for this one as well, perhaps with the help of `getOrElse`.

feiranwang replied:
I'll look into it.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -358,29 +434,37 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
       }
     }
+
+    def generateCanonicalVarFromBodies(bodies: List[Body], indexPrefix: String) {
```
Could you add comment on how this is handling ModifierAtom specially, and why that's necessary?

feiranwang replied:
Sure.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogDeltaDeriver.scala:
```diff
@@ -31,46 +31,51 @@ object DeepDiveLogDeltaDeriver{
+    cq.bodies foreach { bodies =>
+      // We don't support deriving delta rules for modifiers
+      bodies foreach {
+        case x: ModifierAtom => throw new RuntimeException("Deriving delta rules for modifier atom is not supported!")
```
Using `collect` below seems like a good idea to keep the code simple, because there's a check here.  However, if someone happens to add a fourth class, they may get silently ignored by the rest of the code, right?  So, I'd suggest writing this guard in a safer way as below:

``` scala
case _: Atom | _: Cond => // supported
case _ => throw new RuntimeException ...
```

I think you can use `sys.error("...")` for throwing the same exception.

feiranwang replied:
Sounds good.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala:
```diff
@@ -176,13 +184,14 @@ class DeepDiveLogParser extends JavaTokenParsers {
 
   // conditional expressions
   def compareOperator = "LIKE" | ">" | "<" | ">=" | "<=" | "!=" | "=" | "IS" | "IS NOT"
+  def atomModifier = "EXIST" | "OUTER"
```
I'd insist on going with keywords `EXISTS` and `OPTIONAL`.

feiranwang replied:
Sure

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala:
```diff
@@ -18,10 +18,18 @@ case class ConstExpr(value: String) extends Expr
+
+sealed trait Body
+case class Atom(name : String, terms : List[Expr]) extends Body
+case class ModifierAtom(modifier: AtomModifier, bodies: List[Body]) extends Body
```
How about renaming to `ModifiedAtom`?  Code gets confusing, especially when `AtomModifier` is written right next to `ModifierAtom`.

feiranwang replied:
I first called it ModifiedAtom, but I found it may be confusing: it looks like an atom that is modified... maybe we should come up with a better name?

netj replied:
Right, that's true.  How about ModifiedBody?

feiranwang replied:
`ModifiedSomething` sounds like something is modified (changed)... I don't have a good idea right now... will think about it.

netj replied:
How about QuantifiedBody then since they roughly correspond to logical quantifiers?  I also don't like ModifiedSomething...

feiranwang replied:
I think it's good for EXISTS and ALL, but a bit weird for OPTIONAL...

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala:
```diff
@@ -158,11 +165,12 @@ class DeepDiveLogParser extends JavaTokenParsers {
+  // expression
   def expr : Parser[Expr] =
-    ( lexpr ~ operator ~ expr ^^ { case (lhs ~ op ~ rhs) => BinaryOpExpr(lhs, op, rhs) }
+    ( "_" ^^ { case _ => Placeholder() }
```
I wished we first separated patterns from expr before supporting underscore wildcards.  If underscore is used in the head, it doesn't make sense and should be ideally a parse error.

Without complex data type support, the cases for pattern type could be as simple as a wildcard or a binding to a variable or constraint by an expression.  Introducing that may actually make the unification code more clear.  How hard does this sound?  We could do this in a separate PR if it's too much.

feiranwang replied:
I agree it doesn't make sense to put a placeholder in the head. The idea is we use patterns for terms in the body atom, and expressions for head atom terms? 

I think we can remove placeholder in this PR, and do the pattern stuff in a separate PR, since this PR is already big.

netj replied:
Yeah that sounds good.  In that other PR let's revise the grammar clearly to allow only exprs in the head and patts in body.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala:
```diff
@@ -205,20 +214,25 @@ class DeepDiveLogParser extends JavaTokenParsers {
+  def atom = relationName ~ "(" ~ repsep(expr, ",") ~ ")" ^^ {
+    case (r ~ "(" ~ patterns ~ ")") => Atom(r, patterns)
+  }
+  def modifierAtom = (opt("NOT") ~ "EXIST" | "OUTER") ~ "[" ~ rep1sep(cqBody, ",") ~ "]" ^^ { case (m ~ _ ~ b ~ _) =>
```
Perhaps `atomModifier` parser is unused, and this is where should be changed.

Does the user have to write `NOT EXISTS` to emulate `ALL` constraints?  Can we actually expand that syntactic sugar from the parser and support writing `ALL` directly?  Also I thought we were going to use `!` symbol for negation.

feiranwang replied:
Sure I'll add ALL. The idea to transform it into EXISTS form in the parser sounds good.

I'll change it to use ! for negation.

netj replied:
I'm a little worried about introducing negation without putting much thought into it, but that's probably fine since we don't plan to support recursion any time soon.

It may be slightly complicated since you also have to negate any Cond in the nested body (De Morgan's law).  Maybe it's better to keep this out from the parser and have a desugaring step, like the SemanticChecker because I think we want to print as it is when pretty printed.

feiranwang replied:
That makes sense. Simply negating in parser is kind of abusing parser's responsibility, and cause Printer to print different stuff. I think the desugaring step can happen in compilation, as it actually belongs to the compilation process.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogSemanticChecker.scala:
```diff
@@ -45,17 +45,23 @@ object DeepDiveLogSemanticChecker extends DeepDiveLogHandler {
+        s.q.bodies foreach { x =>
+          x foreach {
+            case a: Atom => checkRelation(a.name)
+            case _ =>
```
Shouldn't we recursively descend into modified bodies as well?

feiranwang replied:
Sure

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogSemanticChecker.scala:
```diff
@@ -45,17 +45,23 @@ object DeepDiveLogSemanticChecker extends DeepDiveLogHandler {
       case s: InferenceRule => {
-        s.q.bodies foreach { x => 
-          x foreach { a => checkRelation(a.name) }
+        s.q.bodies foreach { x =>
```
It seems a recursive checker for bodies could be reused here.

feiranwang replied:
I'll look into it.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogSemanticChecker.scala:
```diff
@@ -95,16 +101,21 @@ object DeepDiveLogSemanticChecker extends DeepDiveLogHandler {
     def checkCq(cq: ConjunctiveQuery) {
       checkAtom(cq.head)
-      cq.bodies foreach { x => x foreach checkAtom }
+      cq.bodies foreach { x =>
```
You could define a higher order iterator for bodies that applies the checkers plugged into it to `Atom`s, `ModifiedAtom`s, etc. and reuse it from the places where you do `foreach` over bodies.  Instead of defining multiple iterators, you could have a single iterator that accepts optional checkers for each Body type that defaults to no-op, then visits all nodes in the parse tree.

feiranwang replied:
I'll look into it.

netj commented:
Please see my comments on the diff.  Very excited to see these language constructs becoming real!

feiranwang committed:
Refactoring

feiranwang committed:
Add All modifier

feiranwang commented:
@netj Updated accordingly. Thanks!

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -353,35 +240,226 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
+        alias match {
+          case NoAlias => ""
+          case HeadAlias => s""" AS "${relation}.R${index}.${name}" """
+          case WeightAlias => s"${relation}.R${index}.${name}"
```
Overall, it became much clearer, but I'm still unhappy with this alias part.  I feel like we're almost there untying this ugly knot.  It seems there are several factors that unnecessarily complicate things here:
1. sometimes we want the alias to be prefixed with `AS` and sometimes not.
2. sometimes we don't want the alias at all.
3. only `VarExpr` case considers the `AliasStyle` which seems very odd.
4. each case seems to be quite independent from others, meaning we could split this into several smaller pieces.

Initially, it seemed `compileAlias` could be inlined into `compileExprWithAlias` below.  However, this is called from one extra place when deciding weight tying argument (a non-SQL part) with `WeightAlias`.  The other cases when a variable AliasStyle is passed is via `generateSQLHead`, which is `HeadAlias` only for one out of three cases.

How does the following sound?
1. `compileAlias` could be a simple function that maps to a deterministic alias for every expr, and 
2. `generateSQLHead` could append the aliases with `AS` if necessary.
3. Then `compileExprWithAlias` could be inlined and removed.
4. The `WeightAlias` becomes unnecessary as the caller can just get the alias from `compileAlias`.
5. `HeadAlias`/`NoAlias` becomes a boolean flag for `generateSQLHead`.

feiranwang replied:
Sounds good.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -353,35 +240,226 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
+    val limitStr = optionalClause("LIMIT", cq.limit.getOrElse("").toString)
+
+    s"""FROM ${ generateFromClause(cq.bodies(0), "") }
+        ${ optionalClause("WHERE", generateWhereClause(cq.bodies(0), "")) }${groupbyStr}${limitStr}"""
```
It's still a mystery to me how this works with `bodies(0)`.  It may be a good idea to parameterize that index or redefine the function as `Body => String` instead.  Otherwise, I think we should compile a UNION of all bodies.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -472,17 +542,13 @@ object DeepDiveLogCompiler extends DeepDiveLogHandler {
           }
-          val selectStr = variableCols.mkString(", ")
-          val distinctStr = if (tmpCq.isDistinct) "DISTINCT" else ""
+          val headStr = qc.generateSQLHead(resolveColumnFlag)
```
Inline?

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -600,10 +676,10 @@ object DeepDiveLogCompiler extends DeepDiveLogHandler {
-              val weightVar = w.flatMap(s => ss.resolveColumn(s, fakeCQ, AliasOnly)).mkString(", ")
-              if (weightVar == "") "?"
-              else s"?(${weightVar})"
+              val weightVar = w.map(s => qc.compileAlias(VarExpr(s), WeightAlias)).mkString(", ")
```
We should perhaps change the parser to parse a whole expr for `UnknownFactorWeight` to support variables as well as constant literal strings?  Can we use a column holding arbitrary computed values for the weight ids in DeepDive btw?

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogDeltaDeriver.scala:
```diff
@@ -31,46 +31,51 @@ object DeepDiveLogDeltaDeriver{
+      // We don't support deriving delta rules for modifiers
+      bodies foreach {
+        case _: Cond | _: Atom => // supported
+        case _: QuantifiedBody => throw new RuntimeException("Deriving delta rules for modifier atom is not supported!")
```
I think `: QuantifiedBody` should be removed to make this a pure wildcard.  The whole point of this check is to reject all other cases than Cond and Atom.

feiranwang replied:
Sorry the type should not have been here.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogPrettyPrinter.scala:
```diff
@@ -93,36 +95,52 @@ object DeepDiveLogPrettyPrinter extends DeepDiveLogHandler {
+    b match {
+      case b: Atom => printAtom(b)
+      case b: Cond => printCond(b)
+      case b: QuantifiedBody => printQuantifiedBody(b)
```
It may be easier to use later if they're named all `print` since the parameter type is clear enough.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogPrettyPrinter.scala:
```diff
@@ -93,36 +95,52 @@ object DeepDiveLogPrettyPrinter extends DeepDiveLogHandler {
+    val distinctStr = if (cq.isDistinct) "*" else ""
+    val limitStr = cq.limit match {
+      case Some(s) => s" | ${s}"
+      case None => ""
```
Let's try to avoid matching Option type (Some/None) as that usually clutters code.  How about:

``` scala
val limitStr = cq.limit map { s => s" | ${s}" } getOrElse("")
```

Or:

``` scala
val limitStr = cq.limit map { " | " + _ } getOrElse("")
```

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogPrettyPrinter.scala:
```diff
@@ -143,10 +161,13 @@ object DeepDiveLogPrettyPrinter extends DeepDiveLogHandler {
-      else "\n  semantics = " + stmt.semantics
+    ( stmt.function match {
+        case Some(f) => s"\n  function = ${f}"
+        case None => ""
```
Let's shrink this match as well.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogSemanticChecker.scala:
```diff
@@ -36,26 +36,27 @@ object DeepDiveLogSemanticChecker extends DeepDiveLogHandler {
   }
 
+  // iterate over all atoms contained in the body list and apply the checker
+  def checkBodyAtoms(bodies: List[Body], checker: Atom => Unit) : Unit = {
```
Let's [curry](http://www.scala-lang.org/node/135) this function to make it easier to use:

``` scala
def checkBodyAtoms(checker: Atom => Unit)(bodies: List[Body]) : Unit = {
```

Then later in any real checker that matches cases for statements, etc. can be simplified.  For example `checkRelationDefined` below can become:

``` scala
val check = checkBodyAtoms(checkRelation)
...
stmt match {
  case s: ExtractionRule => s.q.bodies foreach check
  case s: InferenceRule => s.q.bodies foreach check
  ...
}
```

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogSemanticChecker.scala:
```diff
@@ -95,16 +96,39 @@ object DeepDiveLogSemanticChecker extends DeepDiveLogHandler {
+  }
+
+  // check if outer join body contains one atom
+  def checkOuterJoin(stmt: Statement) {
```
Can you add a test for this?  Looks like a nice check which deserves one :)  Perhaps after renaming it to something like `checkUselessOptional`?

netj commented:
Thanks for your changes!  Added a few more comments.

feiranwang committed:
Further refactoring

feiranwang committed:
Allow expressions in weights

feiranwang commented:
@netj Updated accordingly. PTAL. Thanks!

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogPrettyPrinter.scala:
```diff
@@ -78,51 +80,64 @@ object DeepDiveLogPrettyPrinter extends DeepDiveLogHandler {
-          case LogicOperator.OR  => s"[${printCond(lhs)}; ${printCond(rhs)}]"
+  // print a body
+  def print(b: Body) : String = {
+    def printAtom(a: Atom) = {
```
I think having these defined at the top level as `print(a: Atom)`, `print(a: QuantifiedBody)` is better as the direct calls don't need to always go through the extra `match`.

feiranwang replied:
I feel like having one exposed function `print(b: Body)` that takes a body and print it is better than exposing `print` for all subclasses of `Body`. If you think that's better, I can change it.

netj replied:
Let's expose them.  Ideally, the `print(b: Body)` method with boring dispatch should be taken care by scala and be optional, but that's a shortcoming of JVM's ad-hoc polymorphism.

feiranwang replied:
Sure.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogSemanticChecker.scala:
```diff
@@ -34,28 +34,31 @@ object DeepDiveLogSemanticChecker extends DeepDiveLogHandler {
+      if (!(heads contains a.name))
+        error(stmt, s"""relation "${a.name}" is not defined""")
     }
+    def check = checkBodyAtoms(checkRelation)(_)
```
You don't need to add `(_)`, do you?

feiranwang replied:
That will result in a compile error.

```
DeepDiveLogSemanticChecker.scala:56: missing arguments for method checkBodyAtoms in object DeepDiveLogSemanticChecker;
[error] follow this method with `_' if you want to treat it as a partially applied function
[error]     def check = checkBodyAtoms(checkRelation)
```

netj replied:
Sorry, I misunderstood [what scala's "partially applied function" actually is](http://stackoverflow.com/questions/14309501/scala-currying-vs-partially-applied-functions).  It's not currying.  We can change `checkBodyAtoms` slightly, inserting a `=`, to make that scala compile error disappear:

``` scala
def checkBodyAtoms(checker: Atom => Unit) = (bodies: List[Body]) : Unit = {
```

feiranwang replied:
I got the following error

```
DeepDiveLogSemanticChecker.scala:41: ';' expected but '=' found.
[error]   def checkBodyAtoms(checker: Atom => Unit) = (bodies: List[Body]) : Unit = {
[error]                                                                           ^
```

I think if we want to use partially applied function, we need to define the function like previously

``` scala
def checkBodyAtoms(checker: Atom => Unit, bodies: List[Body]) : Unit = {
```

and use it as `checkBodyAtoms(checker, _)`

netj replied:
Sorry, my code example was incomplete.  The original equal should've changed to an arrow:

``` scala
def checkBodyAtoms(checker: Atom => Unit) = (bodies: List[Body]) => {
```

and the types could be moved to the def:

``` scala
def checkBodyAtoms(checker: Atom => Unit): List[Body] => Unit = bodies => {
```

feiranwang replied:
IMO, this syntax looks more complicated than it needs to be... what's the advantage of this one over just the simple function prototype?

netj replied:
We're making this generic iterator slightly complicated to simplify its applications.  As a result you don't need that extra `(_)` from all call sites.  The first flavor without type annotation at the def I think is no more complicated than partially applied function syntax (just two more letters).

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogSemanticChecker.scala:
```diff
@@ -95,16 +98,39 @@ object DeepDiveLogSemanticChecker extends DeepDiveLogHandler {
     def checkCq(cq: ConjunctiveQuery) {
       checkAtom(cq.head)
-      cq.bodies foreach { x => x foreach checkAtom }
+      checkBodyAtoms(checkAtom)(cq.bodies.flatten)
```
How about this?

``` scala
cq.bodies foreach checkBodyAtoms(checkAtom)
```

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogSemanticChecker.scala:
```diff
@@ -95,16 +98,39 @@ object DeepDiveLogSemanticChecker extends DeepDiveLogHandler {
+    }
+    stmt match {
+      case s: ExtractionRule => checkOuterJoinBodies(s.q.bodies.flatten)
+      case s: InferenceRule => checkOuterJoinBodies(s.q.bodies.flatten)
```
I suggest avoiding `flatten` to simply call this type of scanning functions.  Unless you want to retain the flattened list, I think it's better to use iterators such as `foreach` and avoid creating temporary lists.

feiranwang replied:
Sounds good.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala:
```diff
@@ -262,18 +276,19 @@ class DeepDiveLogParser extends JavaTokenParsers {
 
   def constantWeight = floatingPointNumberAsDouble ^^ {   KnownFactorWeight(_) }
-  def unknownWeight  = repsep(variableName, ",")   ^^ { UnknownFactorWeight(_) }
+  def unknownWeight  = repsep(expr, ",")   ^^ { UnknownFactorWeight(_) }
```
If we support expr for `unknownWeight` here, why do we need `constantWeight`?  Isn't it more natural for the compiler to recognize constant expr (with int/float) and decide there?

feiranwang replied:
Sounds good. Looking into it.

netj commented on test/expected-output-test/weights/input.ddl:
```diff
@@ -0,0 +1,9 @@
+weight = x + y.
+
+c(x) :- b(x,y,_)
+weight = "string".
```
Since the parser/compiler supports multiple exprs for weight, how about adding that to this test as well?

feiranwang replied:
Sure.

netj commented:
Thanks for your updates.  Other than the known vs. unknown weight issue and a few minor coding issues, it looks great!  I hope we could get the weight part straightened out now while we're at this.

feiranwang committed:
Refactoring

1. Seperate different types of expressions
2. Update floating point number and integer parsers to resolve parsing ambiguation with dot delimiter
3. Unify factor weight
4. More tests for factor weight

netj commented:
Please see my comments on string escape handling.

feiranwang commented:
Updated string escape and pretty printer.

netj commented:
LGTM
